### PR TITLE
docs: steps to integrate telegraf into kepler

### PR DESCRIPTION
**Pull request by Shobhij**
document covers steps to integrate telegraf into kepler.

### CODE REVIEW

**Comment by rootfs** [comment]
@Shobhij can you sign the DCO?
```
Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
In your local branch, run: git rebase HEAD~4 --signoff
Force push your changes to overwrite the branch: git push --force-with-lease origin main
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+-   Windows 10 or later
+
+-   FreeBSD 12 or later
+
+-   macOS 10.15 Catalina or later
+
+Clone the Telegraf repository:
+```
+root@:~# git clone https://github.com/influxdata/telegraf.git
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:97]: add a new line after the code block

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+Run make build from the source directory
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:99]:
```suggestion

```bash
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+root@:~# cd telegraf
+root@:~# make build
+
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:102]:
```suggestion
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:103]: add a new line after the code block

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+Generate a Telegraf config file
+```
+root@:~# telegraf config > telegraf.conf
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:107]: add a new line after the code block

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+and [IPMI
+sensor](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/ipmi_sensor)
+input
+
+**Output Plugin:** [Prometheus output
+plugin](https://github.com/influxdata/telegraf/tree/master/plugins/outputs/prometheus_client)
+must be enabled in Telegraf config in order to store the metrics in
+Prometheus database .
+
+Below is the sample config that have been used to enable all the
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:124]:
```suggestion
Below is the sample config that was used to enable all the
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+above-mentioned plugins. Although, user can enable any other desired
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:125]:
```suggestion
above-mentioned plugins. Although, the user can enable any other desired
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+plugin by commenting out the respective section.
+
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:128]:
```suggestion
```bash
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+  ## If ipmitool should use a cache
+  ## for me ipmitool runs about 2 to 10 times faster with cache enabled on HP G10 servers (when using ubuntu20.04)
+  ## the cache file may not work well for you if some sensors come up late
+  # use_cache = false
+
+  ## Path to the ipmitools cache file (defaults to OS temp dir)
+  ## The provided path must exist and must be writable
+  # cache_path = ""
+
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:341]: add a new line after the code block

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+Run Telegraf with the plugins defined in config file:
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:343]:
```suggestion

```bash
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+root@:~#./telegraf --config telegraf.conf
+```
+
+Download and start the Prometheus container
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:347]: fixup this top level title to be a level 2 or 3 header

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+==========================================================
+
+Prometheus can be installed on a system in various ways. Here it is
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:350]:
```suggestion
Prometheus can be installed on a system in various ways. Here it is simply run inside a container.
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+downloaded and installed as a container.
+
+Create a Prometheus configuration file that is scrapping from both
+Kepler and Telegraf instance:
+
+Sample Prometheus configuration file is as follows:
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:357]:
```suggestion

```bash
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+scrape_configs:
+# The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
+	- job_name: 'kepler'
+	  static_configs:
+		- targets: ['xx.xx.xx:9102'] 
+	- job_name: 'telegraf'
+	  static_configs:
+		- targets: ['xx.xx.xx:9273']
+
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:375]: Add a new line after a code block

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 15]
```diff
@@ -0,0 +1,467 @@
...
+Run the Prometheus container with the created Prometheus configuration file:
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:377]:
```suggestion

```bash
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 16]
```diff
@@ -0,0 +1,467 @@
...
+root@:~# docker run -d -p 9090:9090 -v $PWD/prometheus.yaml:/etc/prometheus/prometheus.yml prom/prometheus
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:379]: Add a new line after a code block

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+On the Prometheus GUI at localhost:9090, it can be confirmed that Prometheus is scrapping from Kepler and Telegraf.
+
+![](../fig/Kepler-Telegraf-Prometheus.png)
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:382]: Needs an image name to pass the markdown linter

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 18]
```diff
@@ -0,0 +1,467 @@
...
+
+Download and start Grafana container
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:384]: Fixup title level

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 19]
```diff
@@ -0,0 +1,467 @@
...
+==================================================
+
+Like, Prometheus, Grafana can be installed on the system in various
+ways. Here, we are installing Grafana's container image.
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:389]:
```suggestion

```bash
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 20]
```diff
@@ -0,0 +1,467 @@
...
+root@:~# docker run -d --network host --name grafana grafana/grafana
+```
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:391]: Add a new line after a code block

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 21]
```diff
@@ -0,0 +1,467 @@
...
+Once Grafana container is running access the Grafana GUI at localhost:3000. Login with default credentials. After login, The Prometheus database needs to be added as a data source into Grafana GUI. Click on “DATA SOURCES” -> “Add your first data source”
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:392]:
```suggestion
Once the Grafana container is running, access the Grafana GUI at localhost:3000. Login with the default credentials. After login, the Prometheus database needs to be added as a data source into the Grafana GUI. Click on `DATA SOURCES` -> `Add your first data source`
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 22]
```diff
@@ -0,0 +1,467 @@
...
+And select Prometheus - > Click “Save and Test”
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:393]:
```suggestion
and select Prometheus - > Click “Save and Test”
```

Comment (maryamtahhan) [reply docs/installation/Kepler-Telegraf-integration-steps.md:393]: should probably add this to the line above it

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 23]
```diff
@@ -0,0 +1,467 @@
...
+
+Dashboard
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:395]: fixup title level

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 24]
```diff
@@ -0,0 +1,467 @@
...
+
+Once Prometheus has been added as a data source, create a dashboard by
+exporting
+[Kepler-Exporter.json](https://github.com/sustainable-computing-io/kepler/blob/main/grafana-dashboards/Kepler-Exporter.json),
+a default Grafana dashboard for Kepler.
+
+This default Kepler dashboard can be edited in order to display Telegraf
+metric along with Kepler metrics.
+
+For example in below shown example, right hand shows Power related
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:406]:
```suggestion
For example in the image below, the right hand panel shows power related
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 25]
```diff
@@ -0,0 +1,467 @@
...
+metrics collected by Telegraf whereas left hand shows Power related
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:407]:
```suggestion
metrics collected by Telegraf whereas left hand panel shows power related
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 26]
```diff
@@ -0,0 +1,467 @@
...
+metrics by Kepler per namespace:
+
+![](../fig/Kepler-Telegraf-dashboard.png)
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:410]: image needs a title

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 27]
```diff
@@ -0,0 +1,467 @@
...
+describes the total energy spent in DRAM by a container.
+
+**Other-\>** Represents kepler\_container\_other\_joules\_total metric
+measures the cumulative energy consumption on other host components
+besides the CPU and DRAM. Generally, this metric is the host energy
+consumption (from acpi) less the RAPL Package and DRAM.
+
+**On Telegraf side:**
+
+
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:429]: Too many new lines

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 28]
```diff
@@ -0,0 +1,467 @@
...
+This default Kepler dashboard can be edited in order to display Telegraf
+metric along with Kepler metrics.
+
+For example in below shown example, right hand shows Power related
+metrics collected by Telegraf whereas left hand shows Power related
+metrics by Kepler per namespace:
+
+![](../fig/Kepler-Telegraf-dashboard.png)
+
+**On Kepler side:**
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:412]: would fix this as a lower level header with `####`

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 29]
```diff
@@ -0,0 +1,467 @@
...
+
+**DRAM-\>** Represents kepler\_container\_dram\_joules\_total metric which
+describes the total energy spent in DRAM by a container.
+
+**Other-\>** Represents kepler\_container\_other\_joules\_total metric
+measures the cumulative energy consumption on other host components
+besides the CPU and DRAM. Generally, this metric is the host energy
+consumption (from acpi) less the RAPL Package and DRAM.
+
+**On Telegraf side:**
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:427]: would fix this as a lower level header with ####

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 30]
```diff
@@ -0,0 +1,467 @@
...
+metric along with Kepler metrics.
+
+For example in below shown example, right hand shows Power related
+metrics collected by Telegraf whereas left hand shows Power related
+metrics by Kepler per namespace:
+
+![](../fig/Kepler-Telegraf-dashboard.png)
+
+**On Kepler side:**
+
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:413]: Maybe create a bullet list for the items below

File: docs/installation/Kepler-Telegraf-integration-steps.md (hunk 27 above)
Comment (maryamtahhan) [reply docs/installation/Kepler-Telegraf-integration-steps.md:429]: Also maybe a bullet list for the items below would make it a little more readable.

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 31]
```diff
@@ -0,0 +1,467 @@
...
+**PKG-\>** Represents kepler\_container\_package\_joules\_total metrics
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:414]:
```suggestion
**PKG-\>** Represents `kepler_container_package_joules_total metrics`
```

You don't need to escape underscores in MD

Comment (maryamtahhan) [reply docs/installation/Kepler-Telegraf-integration-steps.md:414]: same comment goes for all the escaped underscores that follow

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 32]
```diff
@@ -0,0 +1,467 @@
...
+including all cores and uncore components (e.g. last-level cache,
+integrated GPU and memory controller).
+
+**DRAM-\>** Represents kepler\_container\_dram\_joules\_total metric which
+describes the total energy spent in DRAM by a container.
+
+**Other-\>** Represents kepler\_container\_other\_joules\_total metric
+measures the cumulative energy consumption on other host components
+besides the CPU and DRAM. Generally, this metric is the host energy
+consumption (from acpi) less the RAPL Package and DRAM.
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:425]:
```suggestion
consumption (from ACPI) less the RAPL Package and DRAM.
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 33]
```diff
@@ -0,0 +1,467 @@
...
+describes maximum Thermal Design Power (TDP) available for processor
+package. On Grafana it is the sum of the metrics on both the sockets
+i.e. powerstat\_package\_current\_thermal\
+\_power\_consumptions of socket 0 +
+powerstat\_package\_current\_thermal\_power\_consumptions of socket 1.
+
+**Total DRAM Power metrics number on Kepler side and Telegraf side
+aligns with each other(approximately).**
+
+Telegraf- IPMI metric
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:452]:
```suggestion
Telegraf-IPMI metric
```

Also fixup header level with the appropriate number of `##`

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 34]
```diff
@@ -0,0 +1,467 @@
...
+---------------------
+
+On Kepler dashboard, we are also pulling IPMI metrics which show Power
+Supply Current out %.
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:456]:
```suggestion
supply current out percentage.
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 35]
```diff
@@ -0,0 +1,467 @@
...
+\_power\_consumptions of socket 0 +
+powerstat\_package\_current\_thermal\_power\_consumptions of socket 1.
+
+**Total DRAM Power metrics number on Kepler side and Telegraf side
+aligns with each other(approximately).**
+
+Telegraf- IPMI metric
+---------------------
+
+On Kepler dashboard, we are also pulling IPMI metrics which show Power
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:455]:
```suggestion
On Kepler dashboard, we are also pulling IPMI metrics which show power
```

File: docs/installation/Kepler-Telegraf-integration-steps.md [hunk 36]
```diff
@@ -0,0 +1,467 @@
...
+Supply Current out %.
+
+![](../fig/Kepler-Telegraf-IPMI.png)
```
Comment (maryamtahhan) [inline docs/installation/Kepler-Telegraf-integration-steps.md:458]: figure needs a name for the MD linter

**Review by maryamtahhan — CHANGES_REQUESTED** [review]
Sorry about all the comments - these are mostly markdown linter changes. 
I've proposed adding the linter to the PR workflows and this PR will fail if it's run through the linter.

**Comment by Shobhij** [comment]
> @Shobhij can you sign the DCO?
> 
> ```
> Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
> In your local branch, run: git rebase HEAD~4 --signoff
> Force push your changes to overwrite the branch: git push --force-with-lease origin main
> ```

Done!!

**Comment by Shobhij** [comment]
Looks like by mistake my git rebase HEAD~41 -sign off command, signed off last 41 commits. Can anyone suggest how can I fix this? I was trying to signoff the commits on my pull request and by mistake did HEAD~41..

**Pull request closed without merging**